### PR TITLE
Implement PostgreSQL filters in folds

### PR DESCRIPTION
**Pull request by chewselene**
Adding filtering ability of PostgreSQL as a WHERE clause on the folded subquery.

### CODE REVIEW

File: graphql_compiler/compiler/emit_sql.py [hunk 1]
```diff
@@ -804,16 +831,19 @@ def start_global_operations(self):
 
     def filter(self, predicate):
         """Execute a Filter Block."""
+        # If there is an active fold, add the filter to the current fold.
         if self._current_fold is not None:
-            raise NotImplementedError('Non-_x_count filters inside a fold are not implemented yet.')
+            self._current_fold.add_filter(predicate, self._aliases)
```
[the remark this answers is not in the document]
Comment (chewselene) [reply graphql_compiler/compiler/emit_sql.py]: Let me know if you think the comments are clear. Basically `_x_count` filters are applied in the `WHERE` clause of the whole query, not the folded subquery.

File: graphql_compiler/compiler/emit_sql.py [hunk 2]
```diff
@@ -804,16 +831,19 @@ def start_global_operations(self):
...
-        if self._is_in_optional_scope():
-            sql_expression = sqlalchemy.or_(sql_expression,
-                                            self._came_from[self._current_alias].is_(None))
-        self._filters.append(sql_expression)
+        # Otherwise, add the filter to the compilation state.
+        else:
+            sql_expression = predicate.to_sql(self._dialect,
+                                              self._aliases,
+                                              self._current_alias)
+            if self._is_in_optional_scope():
```
[the remark this answers is not in the document]
Comment (chewselene) [reply graphql_compiler/compiler/emit_sql.py]: The comment explaining this is in the `_join_to_parent_location` function, I believe. @bojanserafimov correct me if I'm wrong.

[the remark this answers is not in the document]
Comment (obi1kenobi) [reply graphql_compiler/compiler/emit_sql.py]: I'm not sure if that applies here 😕@bojanserafimov thoughts?

File: graphql_compiler/compiler/emit_sql.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (obi1kenobi) [reply graphql_compiler/compiler/emit_sql.py]: So if I'm understanding this correctly, it's illegal to call the `filter()` function where the `predicate` references `_x_count` in any way?

[the remark this answers is not in the document]
Comment (chewselene) [reply graphql_compiler/compiler/emit_sql.py]: No, you still call the `filter()` function if there is an `_x_count` it just takes the `else` block of this function.

For filtering on `_x_count` the IR looks something like:
```
...  
Fold(...)
MarkLocation(...)
Unfold()
...
GlobalOperationsStart()
Filter(BinaryComposition(<operation>, FoldedContextField(..., field=_x_count),...)
...
```

vs a filter on a folded local field:

```
...  
Fold(...)
Filter(BinaryComposition(<operation>, LocalField(<local field name>, ...), ...)
MarkLocation(...)
Unfold()
...
```

If we're filtering on a local field, that happens between a `Fold` and `Unfold` block. This is added to the `SQLFoldObject` and added as a `WHERE` clause in the folded subquery (taking the `if self._current_fold is not None` block).

If we're filtering on `_x_count`, this happens after the `GlobalOperationsStart` block and is treated as a global filter, applied to the global `WHERE` clause (taking the `else` block).

[the remark this answers is not in the document]
Comment (obi1kenobi) [reply graphql_compiler/compiler/emit_sql.py]: Got it, that makes sense. Thanks!

File: graphql_compiler/compiler/emit_sql.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bojanserafimov) [reply graphql_compiler/compiler/emit_sql.py]: It looks correct. https://github.com/kensho-technologies/graphql-compiler#constraints-and-rules

**Pull request merged**